### PR TITLE
baseUriParameter on \Raml\Method

### PR DESCRIPTION
**Pull request by CozyLife**
I'm not sure why so many changes show up, but the changes I made are in src/Method.php to fix issue #61.

### CODE REVIEW

**Comment by alecsammon** [comment]
Hmm - did you run some code autoformatter over the code base?

**Comment by CozyLife** [comment]
Nope, but I reviewed your changes on your last commit, and it seems to be the same changes you made. I'm not sure how they appear differently between the repos, though. Perhaps it's going off of my commits to my own repo since the fork? I resynced the repo data to match yours before I made my changes, so my repo got those updates with the latest commit.

**Comment by alecsammon** [comment]
Strange - can you run 
```
./vendor/bin/phpcs --standard=PSR1,PSR2 src
```

this should give no warnings - on master it is ok

**Comment by CozyLife** [comment]
I'm running PHP 5.6 on my dev rig, so I'm not sure it that would work. As for the PSR standards, the formatting isn't the issue. For instance, you simplified my explicit empty() by switching the order of the if/then block to make it implicit. Those changes are in the pull request for some reason.

**Comment by CozyLife** [comment]
Alright, that's weird. You haven't updated your repo lately, yet there are merge conflicts? I'm not sure why, but there have been quite a few issues with keeping everything synchronized between our repos lately. Is there a way to tell GitHub to pull everything fresh from your repo, and then I can just go and push the new changes to mine? I don't want to delete and recreate my fork since it looses my association with the project. I'll look into the issue, too.

As for my latest commit, I added an XML parser as the [RAML spec says](http://raml.org/spec.html#schema) that all parsers must be capable of parsing XML and JSON.

**Comment by CozyLife** [comment]
Apparently to pull from a repo to one you own, you still use a pull request. I've solved all of the issues, and now they're ready for you to pull.

**Comment by CozyLife** [comment]
Are you still waiting on something from me before you pull my code?

**Comment by alecsammon** [comment]
Hi - sorry. Not had much time recently.

firstly can you fix the formatting of the code
```
 ./vendor/bin/phpcs --standard=PSR1,PSR2 src
```
this should produce no errors - i.e. no output.

The baseUriParameter change looks good, but need to understand to investigate the XML stuff more. The original intention with this library was not to include any XML schema parser, I was concerned that the library would start growing to big, and instead it was felt that it would be best just allow user of the library to pass in their own schema parsers into the constructor. However, your code at first glance looks good, so may be good to include.

Also - if you could split your pull requests up into separate features next time that would make my life much easier. The best way of doing this would be to create a branch on your github clone for each feature. You can then create a separate pull request for each branch. You can still merge those branches back into your master, so that you can use the all your changes together.

**Comment by alecsammon** [comment]
Oh - regarding the code formatting. This library uses PSR-2 https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md. I'm sorry - it really should be in the README. Not sure which IDE you use, but you should be able to find a autoformatter.

**Comment by CozyLife** [comment]
That's going to be quite a lot of work on my part to branch it out like that, especially if I make the changes simultaneously and then wind up needing to modify my work after a commit. Sorry, but I'm going to have to decline that request. 

As for the PSR standards, I follow them and Eclipse is set to use them, too. I looked up the code sniffer, and it's for PHP 5.4. I'm on 5.6, so the binary has dependency conflicts if I try to install it. As I already follow the standards, it's a moot point.

**Comment by alecsammon** [comment]
These are the current PSR2 issues on this branch.

```
FILE: /Users/asammon/Projects/personal/php-raml-parser/src/Method.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
 470 | ERROR | [x] Expected 0 spaces after opening bracket; newline
     |       |     found
 471 | ERROR | [x] Spaces must be used to indent lines; tabs are not
     |       |     allowed
 472 | ERROR | [x] Spaces must be used to indent lines; tabs are not
     |       |     allowed
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...al/php-raml-parser/src/Schema/Definition/XmlSchemaDefinition.php
----------------------------------------------------------------------
FOUND 22 ERRORS AFFECTING 20 LINES
----------------------------------------------------------------------
 44 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 45 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 46 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 46 | ERROR | [x] Expected 1 newline after opening brace; 2 found
 47 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 48 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 49 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 50 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 51 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 52 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 53 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 54 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 55 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 56 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 57 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 57 | ERROR | [x] Expected 1 newline after opening brace; 2 found
 58 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 59 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 63 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 64 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 65 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 91 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
----------------------------------------------------------------------
PHPCBF CAN FIX THE 22 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: .../personal/php-raml-parser/src/Schema/Parser/JsonSchemaParser.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 21 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...s/personal/php-raml-parser/src/Schema/Parser/XmlSchemaParser.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 18 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
 19 | ERROR | [x] Spaces must be used to indent lines; tabs are not
    |       |     allowed
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
```

**Comment by alecsammon** [comment]
I've also just upgraded to php 5.6.7 - and have no problems with the code sniffer.

**Comment by CozyLife** [comment]
Yep, tabs are about the only thing I changed from the spec. It's been a while since I read them. To keep things consistent in the code, they should really be spaces.

Does the sniffer change the format? Eclipse messes stuff up when I run the auto formatter. I'll try again today to get the sniffer.

**Comment by CozyLife** [comment]
I just saw the lines about PHPCBF. I'll look into using that.

**Comment by CozyLife** [comment]
All better.

**Comment by brendo** [comment]
> All better.

Kinda, there's no test coverage for any of the new features this PR contains. 

@alecsammon is on the right track. If the pull request is going to contain bug fixes and new features, but no tests to support either of these changes, then it's going to take additional time to merge as each are validated. While I know it's annoying to switch between different branches when you're on a roll, it really is the best way to contribute. 

At the moment there's no flexibility in this PR to actually release the bug fixes without also releasing the new feature. Everything will have to be shipped together which is unfortunate when trying to adhere to [semver](http://semver.org/) standards.

File: src/Schema/Definition/XmlSchemaDefinition.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    // ---
+
+    /**
+     * Returns the XML Schema as an array
+     *
+     * Credit: @link http://php.net/manual/en/book.simplexml.php#105330
+     *
+     * @return array
+     */
+    public function getXmlArray()
```
Comment (alecsammon) [inline src/Schema/Definition/XmlSchemaDefinition.php]: Hi @CozyLife. Not sure if this does what you think it should do!

Given the following schema
```
<xs:schema attributeFormDefault="unqualified"
elementFormDefault="qualified"
xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="api-request">
        <xs:complexType>
            <xs:sequence>
                <xs:element type="xs:string" name="input"/>
            </xs:sequence>
        </xs:complexType>
    </xs:element>
</xs:schema>
```

this returns the following array
````
'@attributes' => [
    'attributeFormDefault' => 'unqualified',
    'elementFormDefault' => 'qualified'
]
```

Doesn't seam correct to me. What should this function be returning? Or have I misunderstood XML Schema?

**Comment by CozyLife** [comment]
@brendo You have a point with the SemVer comment. I wasn't thinking about that, and I made the mistake of thinking that branching it out in this case was more to lessen the workload for each pull. (I don't mind if it takes a while longer, @alecsammon and the other contributors are doing an awesome job for free, so IMO they can take as long as they'd like to.)

As for the tests, I'll stop being a lazy arse and play with PHPUnit. My framework uses PHPUnit in a slightly different way, so I put off having to manually test with PHPUnit. In hindsight it was a rude thing for me to do, as it's my job as a developer to write tests for my own code, not @alecsammon. I've been swamped with work, but that's no excuse. So, I apologize, and I'll write those today.

@alecsammon I'll test the issue today when I make the testing library, and I'll see why it stopped working. The example you supplied is the same one I used from the RAML spec, and by all rights it should still work. If the method wasn't directly accessed by the parser, or used in validation, then it might have gotten missed. Just one more reason to not skip the tests to speed things up. ;)

**Comment by alecsammon** [comment]
Hi @CozyLife. I've already added unit tests for the base uri parameters part - so don't worry about those ones (if you've not already started). 

I've started writing tests for the XML schema stuff - that's how I found the issue with getXmlArray(). It's all on a branch on my repo if you want to take a look. (I have broken some things, so not ready yet).

And thank you for your contributions, it's much easier for me to merge if there are unit tests, and should make merges quicker, but it's not essential.

**Comment by alecsammon** [comment]
Sorry - just to confirm. Please do continue to push code, in whatever way is easiest for you. Something is always always better than nothing. Don't worry if pull requests get big, or if there are no tests.

If things are spit into separate pull request with tests then I can get them merged much quicker. But I'm always happy to add tests if needed - may just take me some more time....

**Comment by CozyLife** [comment]
I see that we think alike, @alecsammon. My tests are very similar to yours, and I've added a separate fixture to keep things tidy. My tests are almost finished, so you don't need to keep working on those. I've found that the array stuff is all wrong. Thanks for catching that. I'm in the process of working that out now.

As for the branching, I agree with the SemVer standard. Although my repo is just a personal fork where no one else will be effected if I don't version it, it does make a difference in how your script can exhibit proper versioning. Therefore, I'll stick with the SemVer standard so that any software that depends on it will stay stable.

**Comment by CozyLife** [comment]
I accidentally gave it an internally used commit message instead of a proper one. Oh well, everything is ready to go. The tests succeed, but I removed the conversion to an array since I wasn't using it, and it was starting to bloat the code for a feature no one was using yet. I also ran PHPCBF on all of the code, so you'll see some formatting changes on other files. I tested it after running the formatter.

**Comment by brendo** [comment]
Thanks @CozyLife, I wasn't sure how to bring it up but I can see we're all on the same page here so things are gonna be just fine. Many thanks for your contributions :)

**Pull request merged**